shared/prow/prow.go

```diff
@@ -0,0 +1,336 @@
+		return 0, err
+	}
+	latestBuild, err := strconv.Atoi(strings.TrimSuffix(string(contents), "\n"))
+	if err != nil {
+		return 0, err
+	}
+
+	return latestBuild, nil
+}
+
+// GetBuild gets build struct based on job info
+// 	No gcs operation
+func (j *Job) GetBuild(buildID int) *Build {
+	return &Build{
+		Bucket: BucketName,
+		JobName: j.Name,
+		JobType: j.Type,
+		BuildID: buildID,
+	}
+}
+
+// GetFinishedBuilds gets all builds that have finished,
+// by looking at existence of "finished.json" file
+func (j *Job) GetFinishedBuilds() []Build {
+	var finishedBuilds []Build
+	builds := j.GetBuilds()
+	for _, build := range builds {
+		if true == build.IsFinished() {
+			finishedBuilds = append(finishedBuilds, build)
+		}
+	}
+	return finishedBuilds
+}
+
+// GetBuilds gets all builds from this job on gcs
+func (j *Job) GetBuilds() []Build {
+	var builds []Build
+	gcsBuildPaths := gcs.ListDirectChildren(ctx, j.Bucket, j.GetPath())
+	for _, gcsBuildPath := range gcsBuildPaths {
+		buildID, err := GetBuildIDFromBuildPath(gcsBuildPath)
+		if nil != err { // this last part of gcs path is not a valid int64, should not be a build
+			continue
+		}
+		build := Build{
+			Bucket: j.Bucket,
+			JobName: j.Name,
+			BuildID: buildID,
+		}
+		if true == build.IsStarted() {
+			builds = append(builds, build)
+		}
+	}
+	return builds
+}
+
+// GetLatestBuilds get latest builds from gcs
+// The timestamp of gcs directories are not usable, 
+// as they are all set to '0001-01-01 00:00:00 +0000 UTC',
+// so use 'started.json' creation date for latest builds
+func (j *Job) GetLatestBuilds(count int) []Build {
+	builds := j.GetFinishedBuilds()
+	sort.Slice(builds, func(i, j int) bool {
+		startedTime1, err1 := builds[i].GetStartedTime()
+		if nil != err1 {
+			return false
+		}
+		startedTime2, err2 := builds[j].GetStartedTime()
+		if nil != err2 {
+			return true
+		}
+		return startedTime1 > startedTime2
+	})
+	if len(builds) < count {
+		return builds
+	}
+	return builds[:count]
+}
+
+// IsStarted check if build has started by looking at "started.json" file
+func (b *Build) IsStarted() bool {
+	return gcs.Exist(ctx, BucketName, path.Join(b.GetPath(), StartedJSON))
+}
+
+// IsFinished check if build has finished by looking at "finished.json" file
+func (b *Build) IsFinished() bool {
+	return gcs.Exist(ctx, BucketName, path.Join(b.GetPath(), FinishedJSON))
+}
+
+// GetStartedTime gets started timestamp of a build,
+// return -1 means failed to get timestamp
+func (b *Build) GetStartedTime() (int64, error) {
+	gcsStartedJSONPath := path.Join(b.GetPath(), StartedJSON)
+	contents, err := gcs.Read(ctx, b.Bucket, gcsStartedJSONPath)
+	if nil != err {
+		return -1, err
+	}
+	var started Started
+	if err = xml.Unmarshal(contents, &started); nil != err {
+		return -1, err
+	}
+
+	return started.Timestamp, nil
+}
+
+// GetFinishedTime gets finished timestamp of a build,
+// return -1 means failed to get timestamp
+func (b *Build) GetFinishedTime() (int64, error) {
```

Comment:
`GetStartedTime` and `GetFinishedTime` are basically the same, can you combine the common code in a helper function?

Reply:
good idea, done